lib/opencensus/trace/exporters/stackdriver/converter.rb

```diff
@@ -0,0 +1,354 @@
+# Copyright 2018 OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "google/devtools/cloudtrace/v2/trace_pb"
+require 'google/protobuf/well_known_types'
+
+module OpenCensus
+  module Trace
+    module Exporters
+      class Stackdriver
+        ##
+        # An object that converts OpenCensus span data objects to Stackdriver
+        # Trace V2 protos.
+        #
+        # You should use one converter instance to convert the spans for a
+        # single export request, because the converter will keep track of and
+        # omit duplicate stack traces. Use a new converter instance for the
+        # next request.
+        #
+        # @private
+        #
+        class Converter
+          ##
+          # @private
+          # Alias for the V2 Cloudtrace protos namespace
+          #
+          TraceProtos = Google::Devtools::Cloudtrace::V2
+
+          ##
+          # Create a converter
+          #
+          def initialize
+            @stack_trace_hash_ids = {}
+          end
+
+          # rubocop:disable Metrics/AbcSize
+
+          ##
+          # Convert a span object.
+          #
+          # @param [OpenCensus::Trace::Span] obj OpenCensus span object
+          # @param [String] project_id Google project ID
+          # @return [Google::Devtools::Cloudtrace::V2::Span] The generated
+          #     proto
+          #
+          def convert_span obj, project_id
```

Comment:
Should project_id be an attribute of the converter? The only reason it *might* matter is that the `stack_trace_hash_id` might not be shared across projects in the edge case that an app is reporting across multiple projects (might not be a real case).

As an attribute of the converter, you'd have to instantiate one converter instance per project and it would keep the set of stack trace hash ids separate.